### ISSUE TITLE
Matching format for multi input keras model with tfrecord

### ISSUE THREAD

**Issue by teodor440** [PAIR-code/what-if-tool#77]
I am new to tensorflow and I'm quite confused about the format the wit is expecting from the tfrecords to feed the model
Basically I have a multi input keras model with the following signature:
```
The given SavedModel SignatureDef contains the following input(s):
  inputs['byte_entropy'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 256)
      name: serving_default_byte_entropy:0
  inputs['data_directories'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 30)
      name: serving_default_data_directories:0
  inputs['exports'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 128)
      name: serving_default_exports:0
  inputs['general'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 10)
      name: serving_default_general:0
  inputs['header'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 62)
      name: serving_default_header:0
  inputs['histogram'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 256)
      name: serving_default_histogram:0
  inputs['imports'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 1280)
      name: serving_default_imports:0
  inputs['section'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 255)
      name: serving_default_section:0
  inputs['strings'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 104)
      name: serving_default_strings:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['final_output'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 1)
      name: StatefulPartitionedCall:0
Method name is: tensorflow/serving/predict
```
And some tfrecords (records of Examples) with the same features
```
feature={'histogram': FixedLenFeature(shape=[256], dtype=tf.float32, default_value=None), 
'byte_entropy': FixedLenFeature(shape=[256], dtype=tf.float32, default_value=None), 
'strings': FixedLenFeature(shape=[104], dtype=tf.float32, default_value=None), 
'general': FixedLenFeature(shape=[10], dtype=tf.float32, default_value=None), 
'header': FixedLenFeature(shape=[62], dtype=tf.float32, default_value=None), 
'section': FixedLenFeature(shape=[255], dtype=tf.float32, default_value=None), 
'imports': FixedLenFeature(shape=[1280], dtype=tf.float32, default_value=None), 
'exports': FixedLenFeature(shape=[128], dtype=tf.float32, default_value=None), 
'data_directories': FixedLenFeature(shape=[30], dtype=tf.float32, default_value=None), 
'final_output': FixedLenFeature(shape=(), dtype=tf.float32, default_value=None)}
```
When trying to show it as a regression I get an invalid argument error
![image](https://user-images.githubusercontent.com/14222938/80761895-21d70c80-8b44-11ea-9c8a-9a27c93426ba.png)
```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "input size does not match signature: 1!=9 len({byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}) != len({byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}). Sent extra: {byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}. Missing but required: {byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}."
        debug_error_string = "{"created":"@1588279350.370000000","description":"Error received from peer ipv6:[::1]:8500","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"input size does not match signature: 1!=9 len({byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}) != len({byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}). Sent extra: {byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}. Missing but required: {byte_entropy,data_directories,exports,general,header,histogram,imports,section,strings}.","grpc_status":3}"
```
Where or how should I specify which tfrecord goes to which input in the model?

And If you are kind do you know how may I alter the model prediction to make it suitable for classification? At the moment the prediction is a number from zero to one and to suit the wit it should be an array with two probabilities. I know this is not a question about wit especially

**Comment by tolga-b** [comment]
Hi teodor440,
Just to clarify, is your model being served by tf.Serving or is it loaded locally in a notebook?
If you have local access to your model, then it may be easier to use custom_predict_fn. [Here is a demo](https://colab.sandbox.google.com/github/pair-code/what-if-tool/blob/master/WIT_Toxicity_Text_Model_Comparison.ipynb#scrollTo=nStoYhqT80WH) that shows how to call a keras model with your inputs. WIT in this case will call your custom_predict_fn with tf.Examples and you can unpack them to send it to your keras model.

If your model output is already between 0 and 1 (softmax) you can create an array of [pred, 1-pred] as done [here](https://colab.sandbox.google.com/github/pair-code/what-if-tool/blob/master/WIT_COMPAS_with_SHAP.ipynb#scrollTo=KF00pJvkeicT) (check custom_predict_with_shap function). If your outputs are logits, you can call tf.softmax to normalize it between 0 and 1. It's tricky and task dependent in the case where it's a regression task with multiple output values like house price prediction. You may be able to binarize the task by thresholding the output values (e.g. above 50k, below 50k) and train a binary model on the new thresholded dataset and use that instead.

**Comment by teodor440** [comment]
I was trying to use it with tensorflow serving, initially I was able to use the workaround in the notebook with a custom_predict_fn, although not as efficient as running the tensorboard server
My problem was the format I was providing, I was able to match the format wit was expecting by creating an input_fn with ``tf.estimator.export.build_parsing_serving_input_receiver_fn`` and keeping the input tensor with the name **examples**
Seems like class predictions are labeled ok although the model output is not really of the suggested format for a classifier (outputs only a probability between 0 and 1)
Thank you for your attention
